### PR TITLE
Add more Grpc.Tools integration tests

### PR DESCRIPTION
**Pull request by tonydnewell**
Further Grpc.Tools tests
- TestCharactersInName - proto files with dots and numbers in their names
- TestExtraOptions - test setting AdditionalProtocArguments, OutputOptions and GrpcOutputOptions
- TestGrpcServices - test setting GrpcServices to "none", "client", "server" and "both"
- TestSetOutputDirs - test setting OutputDir and GrpcOutputDir

### CODE REVIEW

**Comment by tonydnewell** [comment]
@jtattermusch for review

File: src/csharp/Grpc.Tools.Tests/scripts/fakeprotoc.py [hunk 1]
```diff
@@ -220,32 +220,43 @@ def _get_cs_files_to_generate(protofile, proto_to_generated):
...
         # if not absolute, it is relative to project directory
-        out_dir = os.path.abspath("%s/%s" % (projectdir, out_dir))
+        grpc_out_dir = os.path.abspath("%s/%s" % (projectdir, grpc_out_dir))
 
-    # Ensure out_dir exists
-    if not os.path.isdir(out_dir):
-        os.makedirs(out_dir)
+    if not os.path.isabs(csharp_out_dir):
+        # if not absolute, it is relative to project directory
+        grpc_out_dir = os.path.abspath("%s/%s" % (projectdir, csharp_out_dir))
```
Comment (jtattermusch) [inline src/csharp/Grpc.Tools.Tests/scripts/fakeprotoc.py:238]: bug here: need to assign to `csharp_out_dir =`?

File: src/csharp/Grpc.Tools.Tests/MsBuildIntegrationTest.cs [hunk 2]
```diff
@@ -126,6 +126,61 @@ public void TestProtoOutsideProject()
...
+            SetUpForTest(nameof(TestExtraOptions));
+
+            var expectedFiles = new ExpectedFilesBuilder();
+            expectedFiles.Add("file.proto", "File.cs", "FileGrpc.cs");
+
+            TryRunMsBuild("TestExtraOptions", expectedFiles.ToString());
+        }
+
+        [Test]
+        public void TestGrpcServices()
```
[the remark this answers is not in the document]
Comment (tonydnewell) [reply src/csharp/Grpc.Tools.Tests/MsBuildIntegrationTest.cs]: Changed to TestGrpcServicesMetadata as it is the GrpcServices metadata on the Protobuf item that is being tested. Not sure if there is a better name.

File: src/csharp/Grpc.Tools.Tests/MsBuildIntegrationTest.cs [hunk 3]
```diff
@@ -126,6 +126,61 @@ public void TestProtoOutsideProject()
...
+            expectedFiles.Add("clientonly.proto", "Clientonly.cs", "ClientonlyGrpc.cs");
+            expectedFiles.Add("clientandserver.proto", "Clientandserver.cs", "ClientandserverGrpc.cs");
+
+            TryRunMsBuild("TestGrpcServices", expectedFiles.ToString());
+        }
+
+        [Test]
+        public void TestSetOutputDirs()
+        {
+            // Test setting different GrpcOututDir and OutputDir
```
[the remark this answers is not in the document]
Comment (tonydnewell) [reply src/csharp/Grpc.Tools.Tests/MsBuildIntegrationTest.cs]: Fixed

File: src/csharp/Grpc.Tools.Tests/scripts/fakeprotoc.py (hunk 1 above)
Comment (tonydnewell) [reply src/csharp/Grpc.Tools.Tests/scripts/fakeprotoc.py]: Fixed

File: src/csharp/Grpc.Tools.Tests/scripts/fakeprotoc.py [hunk 4]
```diff
@@ -220,32 +220,43 @@ def _get_cs_files_to_generate(protofile, proto_to_generated):
...
+    if not os.path.isdir(grpc_out_dir):
+        os.makedirs(grpc_out_dir)
+
+    if not os.path.isdir(csharp_out_dir):
+        os.makedirs(csharp_out_dir)
 
     timestamp = str(datetime.datetime.now())
     for csfile in cs_files_to_generate:
-        csfile_fullpath = "%s/%s" % (out_dir, csfile)
+        if csfile.endswith("Grpc.cs"):
```
[the remark this answers is not in the document]
Comment (tonydnewell) [reply src/csharp/Grpc.Tools.Tests/scripts/fakeprotoc.py:256]: Fixed

File: src/csharp/Grpc.Tools.Tests/scripts/fakeprotoc.py [hunk 5]
```diff
@@ -278,6 +293,16 @@ def _getenv(name):
     return value
 
 
+def _get_last_value_or_none(protoc_arg_dict, name):
```
[the remark this answers is not in the document]
Comment (tonydnewell) [reply src/csharp/Grpc.Tools.Tests/scripts/fakeprotoc.py]: Fixed

**Review by jtattermusch — APPROVED** [review]
LGTM once the tests are green (note to self: check copybara results).

**Pull request merged**